File: 75Review/LinkedList/143.go
package LinkedList

func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	// first thing is to split this half
	slow, fast, start := head, head, head
	var prev *ListNode = nil

	for fast != nil && fast.Next != nil {
		prev = slow
		slow = slow.Next
		fast = fast.Next.Next
	}

	// wherever the left pointer is at, that is the index we cut off
	prev.Next = nil
	secondHalf := reverse(slow)

	for start != nil && secondHalf != nil {
		list1Next := start.Next
		list2Next := secondHalf.Next

		start.Next = secondHalf

		if list1Next != nil {
			secondHalf.Next = list1Next
		}

		start = list1Next
		secondHalf = list2Next
	}
}

func reverse(curr *ListNode) *ListNode {
	var prev *ListNode = nil

	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	return prev
}

File: src/2025/XPrep/143.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

ListNode *reverse(ListNode *head) {
    ListNode *prev = nullptr;
    while (head) {
        auto next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }

    return prev;
}

class Solution {
  public:
    void reorderList(ListNode *head) {
        if (!head || !head->next) return;

        ListNode *slow = head;
        ListNode *fast = head;
        ListNode *prev = nullptr;

        while (fast && fast->next) {
            prev = slow;
            fast = fast->next->next;
            slow = slow->next;
        }

        prev->next = nullptr;
        ListNode *second = reverse(slow);

        auto first = head;
        while (first && second) {
            auto next1 = first->next;
            auto next2 = second->next;

            first->next = second;
            if (next1) second->next = next1;

            first = next1;
            second = next2;
        }
    }
};

// first: 1 2
// second 5 4 3